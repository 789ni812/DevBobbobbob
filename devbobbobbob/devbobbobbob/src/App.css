.App {
  background: #1e1e1e;
  color: #cccccc;
}

a {
  color: #6a9955;
}

a:hover {
  color: #b5cea8;
}

h1,
h2,
h3,
h4,
h5 {
  /* color: #6796e6; */
  color: #4ec9b0;
  /* color: #ce9178; */
}
