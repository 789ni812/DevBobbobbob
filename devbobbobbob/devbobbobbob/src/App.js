function App() {
  return (
    <>
        <h1>DevBobBobBob</h1>
    </>
  );
}

export default App;
