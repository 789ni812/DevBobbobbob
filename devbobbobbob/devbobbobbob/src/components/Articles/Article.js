import React from 'react';
//import PropTypes from 'prop-types';





function Article(props) {


return (
<div data-test="component-article">
<h1 data-test="articleTitle">{props.article.title}</h1>
    <div data-test="articleDescription"></div>
    <div data-test="articleCreatedDate"></div>
    <div data-test="articleAuthor"></div>
    <div data-test="articleTags"></div>
    </div>
)    
};

// TODO Add this prop types checking.
// Article.propTypes = {
//     article: PropTypes.arrayOf(
//         PropTypes.shape({
//             title: PropTypes.string.isRequired
//         })
//     ).isRequired,
// };


export default Article;