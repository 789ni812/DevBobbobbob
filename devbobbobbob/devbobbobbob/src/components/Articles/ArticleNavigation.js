function ArticleNavigation() {
  return <h3>component loaded: Article Navigation </h3>;
}

export default ArticleNavigation;
