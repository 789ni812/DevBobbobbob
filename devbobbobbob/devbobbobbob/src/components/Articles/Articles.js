import React, { useEffect, useState } from "react";
import ArticlesService from "../../services/ArticlesService";

const Articles = () => {
  const [articles, setArticles] = useState([]);
  const [currentArticle, setCurrentArticle] = useState(null);

  useEffect(() => {
    retrieveArticles();
  }, []);

  const retrieveArticles = () => {
    ArticlesService.getAll().then((response) => {
      setArticles(response.data.data);
    });
  };

  const setActiveArticle = (article) => {
    setCurrentArticle(article);
  };

  /**
   * Create a Div tag and then renders the HTML included in the string
   * ref: https://reactjs.org/docs/dom-elements.html
   * //TODO Move this helper function to a utility file and include it.
   * @param {string } htmlToRender
   * @returns {string} rendered HTML wrapped in <div></div>
   */
  const dangourslySetInnerHTMLHelper = (htmlToRender) => {
    return (
      <div
        dangerouslySetInnerHTML={{
          __html: htmlToRender,
        }}
      ></div>
    );
  };

  return (
    <div>
      Articles
      <ul>
        {articles &&
          articles.map((resource) => (
            <li
              className="hyperlink"
              key={resource.id}
              onClick={() => setActiveArticle(resource)}
            >
              {resource.attributes.title}
            </li>
          ))}
      </ul>
      {currentArticle ? (
        <div>
          Current resource selected
          <h2>{currentArticle.attributes.title}</h2>
          {currentArticle.attributes.body ? (
            dangourslySetInnerHTMLHelper(currentArticle.attributes.body.value)
          ) : (
            <div>negative</div>
          )}
          <div>
            {/* 
            //FIXME CORB issue when trying to render an image 
            See https://drupal.stackexchange.com/questions/282092/jsonapi-cross-origin-read-blocking-corb-blocked-cross-origin-response-in-vue
            and https://drupal.stackexchange.com/questions/245903/how-do-i-set-up-cors
            
            */}
            {/* <img
              src={currentArticle.relationships.field_image.links.related.href}
              alt={currentArticle.relationships.field_image.data.meta.alt}
            /> */}
          </div>
          {console.log(
            "CurrentResource: ",
            currentArticle.relationships.field_image.links
          )}
        </div>
      ) : (
        <div>Click an article resource for more info...</div>
      )}
    </div>
  );
};

export default Articles;
