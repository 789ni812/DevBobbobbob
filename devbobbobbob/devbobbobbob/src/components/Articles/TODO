
Quick Fire round:
    [x ] stop article component rendering by default in Articles component
    [x] comment out MArkdown test example - to review later
    
    ## React router
    [x ] Install react-router-dom 
    [ ] Article list to render a link to the actual article
            [ ]  Research how route will work with dynamic article.id
            //https://codesandbox.io/s/react-router-url-parameters-rrbsb?from-embed=&file=/example.js
            [ ] Ensure Props is sent to Article from react-router
    [ ] Article navigation to render a back to articles list
    [ ] Article navigation to render tags as links 
    [ ] Depending on tag click from article navigation will render a page specific to that tag



------
## Next

[x ] Move the state management out of the Article componet and into its parent (articles component)

[] Add following articleArray props to the Article component
        [x] articleTitle
        [x] articleDescription
        [x] articleCreatedDate
        [x] articleAuthor
        [x] articleTags (these are an array)
            [x] Research how to test for arrays 
            (Note used toContain (https://stackoverflow.com/questions/41916612/jest-jasmine-tocontain-fails-even-though-matching-values-are-present))

[ ] When user click on Article links from Articles component - display just that user selected article information
    [x] Create ArticleNavigation component
    [x] include ArticleNAvigation to Articles component
    [x ] ArticleNavigation to render list of articles
    [ ] Ability for user to click on a list item 
    [] wire up react router to manage urls
        [ ] create ArticleSelected state (default to null)
        
    [ ] Clicking article list item triggers the loading of the relevant Article in Article Component
        [ ] create an onClick event that sets the ArticleSelected state to the ID of the article

        [ ] research if a useEffect or similar is required to refresh the update to the state or anything.

        [ ] Enable the Article component to load with information from the articleSelected state









    
-----
## Sometime/Maybe

Article List:
[] Show list of articles
    [] show name
    [] arrange by latest date         
    [] Research how to create unique links to article
    [] Implement unique links for each article

    Article API:
        [] Research how to view content from an headless Drupal 9 platform
        [] Write next steps based on research


------
## DONE

App:
[x] Verify first tests are running correctly
[x] Wire up article components into Articles component
[x] Wire up Article component to App component


Mock Article data:
        [x] Research how to test an array of objects
        [x] Write 3 mock articles as an array of objects

        Add to article object: 
            [x] title
            [x] description
            [x] date
            [x] author
            [x] tags
            [x] publishStatus

    Article Mocked Render:
        [x] create a static article component (only use HTML) to display a full article
            [x] title
            [x] description
            [x] date
            [x] author
            [x] tags
    
    