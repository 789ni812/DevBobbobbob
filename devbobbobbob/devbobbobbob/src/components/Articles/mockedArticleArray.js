export const articlesArray = [
    { 
    title: 'Mocked article array', 
    description: 'This is the description field from the mocked article array.' ,
    createdDate: '08112020',
    author: 'Devbob',
  tags: ['developer','tutorial','react'],
  publishStatus: 'draft'
},
{ 
    title: '2nd Mocked article array', 
    description: 'This is the second mocked article object description' ,
    createdDate: '09112020',
    author: 'spud balls',
  tags: ['developer','quiz','JavaScript'],
  publishStatus: 'published'
},
{ 
    title: 'Three is the number of this article magic or not', 
    description: 'third mocked article object description, I mean we really are on fire here .' ,
    createdDate: '10112020',
    author: 'mcdivers',
  tags: ['design','graphic design','gimp'],
  publishStatus: 'draft'
}
  ];


export default articlesArray;