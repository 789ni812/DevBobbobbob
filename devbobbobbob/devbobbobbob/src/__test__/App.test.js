import React from "react";
import { shallow } from "enzyme";
import App from "../App";
import Articles from "../components/Articles/Articles";
import toJson from "enzyme-to-json";

describe('<App />', () => {
  let wrapper;
  beforeAll(()=>{
      wrapper = shallow(<App />);
  });


it("renders without crashing", () => {
  shallow(<App />);
});

it ('renders App heading', () => {
  const title = (<h1>DevBobBobBob</h1>);
  expect(wrapper.contains(title)).toEqual(true);
});

it('renders the <Articles /> component', ()=>{
  expect(wrapper.containsMatchingElement(<Articles />)).toEqual(true);
});

// it("renders same as snapshot", () => {
//   const tree = shallow(<App />);
//   expect(toJson(tree)).toMatchSnapshot();
// });

});

