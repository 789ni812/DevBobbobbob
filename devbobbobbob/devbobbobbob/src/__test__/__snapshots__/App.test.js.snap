// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> renders same as snapshot 1`] = `
<Fragment>
  <h1>
    DevBobBobBob
  </h1>
</Fragment>
`;
